tiledb/sm/query/readers/aggregators/sum_aggregator.cc

```diff
@@ -0,0 +1,312 @@
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class SumAggregatorStatusException : public StatusException {
+ public:
+  explicit SumAggregatorStatusException(const std::string& message)
+      : StatusException("SumAggregator", message) {
+  }
+};
+
+/** Specialization of safe_sum for int64_t sums. */
+template <>
+void safe_sum<int64_t>(int64_t value, int64_t& sum) {
+  if (sum > 0 && value > 0 &&
+      (sum > (std::numeric_limits<int64_t>::max() - value))) {
+    sum = std::numeric_limits<int64_t>::max();
+    throw std::overflow_error("overflow on sum");
+  }
+
+  if (sum < 0 && value < 0 &&
+      (sum < (std::numeric_limits<int64_t>::min() - value))) {
+    sum = std::numeric_limits<int64_t>::min();
+    throw std::overflow_error("overflow on sum");
+  }
+
+  sum += value;
+}
+
+/** Specialization of safe_sum for uint64_t sums. */
+template <>
+void safe_sum<uint64_t>(uint64_t value, uint64_t& sum) {
+  if (sum > (std::numeric_limits<uint64_t>::max() - value)) {
+    sum = std::numeric_limits<uint64_t>::max();
+    throw std::overflow_error("overflow on sum");
+  }
+
+  sum += value;
+}
+
+/** Specialization of safe_sum for double sums. */
+template <>
+void safe_sum<double>(double value, double& sum) {
+  if ((sum < 0.0) == (value < 0.0) &&
+      (std::abs(sum) >
+       (std::numeric_limits<double>::max() - std::abs(value)))) {
+    sum = sum < 0.0 ? std::numeric_limits<double>::lowest() :
+                      std::numeric_limits<double>::max();
+    throw std::overflow_error("overflow on sum");
+  }
+
+  sum += value;
+}
+
+template <typename T>
+SumAggregator<T>::SumAggregator(
+    const std::string field_name,
+    const bool var_sized,
+    const bool is_nullable,
+    const unsigned cell_val_num)
+    : is_nullable_(is_nullable)
+    , field_name_(field_name)
+    , sum_(0)
+    , validity_value_(0)
+    , sum_overflowed_(false) {
+  if (var_sized) {
+    throw SumAggregatorStatusException(
+        "Sum aggregates must not be requested for var sized attributes.");
+  }
+
+  if (cell_val_num != 1) {
+    throw SumAggregatorStatusException(
+        "Sum aggregates must not be requested for attributes with more than "
+        "one value.");
+  }
+}
+
+template <typename T>
+void SumAggregator<T>::validate_output_buffer(
+    std::string output_field_name,
+    std::unordered_map<std::string, QueryBuffer>& buffers) {
+  if (buffers.count(output_field_name) == 0) {
+    throw SumAggregatorStatusException("Result buffer doesn't exist.");
+  }
+
+  auto& result_buffer = buffers[output_field_name];
+  if (result_buffer.buffer_ == nullptr) {
+    throw SumAggregatorStatusException(
+        "Sum aggregates must have a fixed size buffer.");
+  }
+
+  if (result_buffer.buffer_var_ != nullptr) {
+    throw SumAggregatorStatusException(
+        "Sum aggregates must not have a var buffer.");
+  }
+
+  if (result_buffer.original_buffer_size_ != 8) {
```

Comment:
What if the datatype is smaller than 8 bytes and the caller allocates `sizeof(T)`?

Reply:
We always sum up in a 64bit value to prevent overflows as much as possible.

Reply:
Then shouldn't this check happen when the buffer is set? (https://github.com/TileDB-Inc/TileDB/pull/4113/files#diff-e19b0b9e093512958297fb53f9e1ac2e0f8ac7320c0ce5ea19794a67e3acc4f8R984)

Reply:
Yes and no. If the user uses the API, yes. If this comes from a REST query, we might not check the size when we deserialize.

Reply:
That means that there won't be any error until the query is submitted, and in the case of REST queries: until it hits REST and errors out.